pubsub/stanpubsub/stan.go

```diff
@@ -0,0 +1,645 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package stanpubsub provides a pubsub implementation for NATS Streaming client.
+// See https://docs.nats.io/nats-streaming-concepts/intro for an overview.
+//
+// Use OpenTopic to construct a *pubsub.Topic, and/or OpenSubscription to construct a
+// *pubsub.Subscription. This package uses gob to encode and decode driver.Message to
+// []byte.
+//
+// URLs
+//
+// For pubsub.OpenTopic and pubsub.OpenSubscription, stanpubsub registers
+// for the scheme "stan".
+// The default URL opener will connect to a default server based on the environment variables:
+// 	- STAN_SERVER_URL	- url address of the NATS server
+//	- STAN_CLUSTER_ID	- NATS Streaming cluster id
+//	- STAN_CLIENT_ID	- NAST Streaming client id
+// To customize the URL opener, or for more details on the URL format,
+// see URLOpener.
+// See https://gocloud.dev/concepts/urls/ for background information.
+//
+// Message Delivery Semantics
+//
+// NATS Streaming supports at-least-once semantics; by default applications don't need to call Message.Ack,
+// and must not call Message.Nack.
+// Optionally a user can set a subscription to be manually acked by using `stan.SetManualAckMode` option
+// or by setting `manualAck` query parameter when setting from URL.
+// See https://godoc.org/gocloud.dev/pubsub#hdr-At_most_once_and_At_least_once_Delivery
+// for more background.
+//
+// As
+//
+// stanpubsub exposes the following types for As:
+//  - Topic: stan.Conn
+//  - Subscription: stan.Subscription
+//  - Message.BeforeSend: None.
+//  - Message.AfterSend: None.
+//  - Message: *stan.Msg
+
+package stanpubsub
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"path"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/stan.go"
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/batcher"
+	"gocloud.dev/pubsub/driver"
+)
+
+func init() {
+	dd := new(defaultDialer)
+	pubsub.DefaultURLMux().RegisterTopic(Scheme, dd)
+	pubsub.DefaultURLMux().RegisterSubscription(Scheme, dd)
+}
+
+func connect() (*URLOpener, error) {
+	serverURL := os.Getenv("STAN_SERVER_URL")
+	if serverURL == "" {
+		return nil, errors.New("STAN_SERVER_URL environment variable not set")
+	}
+	clusterID := os.Getenv("STAN_CLUSTER_ID")
+	if clusterID == "" {
+		return nil, errors.New("STAN_CLUSTER_ID environment variable not set")
+	}
+	clientID := os.Getenv("STAN_CLIENT_ID")
+	if clientID == "" {
+		return nil, errors.New("STAN_CLIENT_ID environment variable not set")
+	}
+	conn, err := stan.Connect(clusterID, clientID, stan.NatsURL(serverURL))
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial STAN_SERVER_URL %q: STAN_CLUSTER_ID: %q STAN_CLIENT_ID: %q,  %v", serverURL, clusterID, clientID, err)
+	}
+	return &URLOpener{Connection: conn, SubscriptionOptions: DefaultSubscriptionOptions}, nil
```

Comment:
Let's leave the `SubscriptionOptions` out. The opener should deal with the default value explicitly, since someone could create an `URLOpener` manually as well.